# Data Mining - Insurance Project
**Group Members**:
- Canhoto, Vera (m20210659@novaims.unl.pt)
- Correa, Lucas (m20211006@novaims.unl.pt)
- Shin, Doyun (m20200565@novaims.unl.pt)

In [1]:
# Jupyter notebook display size adjustment
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:95% !important; }</style>"))

<a class="anchor" id="0.1"></a>

# **Table of Contents**

Structured as per CRISP-DM method, but it was used flexibly to better reflect the focus of the project. Further qualitative elements are addressed in the report attached where applicable.

1. [Business Understanding](#1)
   - 1.1 [Business Objectives](#1.1)
   - 1.2 [Situation Assessment](#1.2)
   - 1.3 [Data Mining Goals](#1.3)
   - 1.4 [Project Plan](#1.4)


2. [Data Understanding](#2)
   - 2.1 [Initial Data Collection](#2.1)
   - 2.2 [Data Description](#2.2)
   - 2.3 [Data Exploration](#2.3)
   - 2.4 [Data Quality Verification](#2.4)
  
  
3. [Data Preparation](#3)
   - 3.1 [Data Selection](#3.1)
   - 3.2 [Data Cleaning](#3.2) 
   - 3.3 [Data Construction](#3.3) 
   - 3.4 [Data Integration](#3.4) 
   - 3.5 [Data Formatting](#3.5)
   - 3.6 [Dataset Description](#3.6)
   
   
4. [Modeling](#4)
   - 4.1 [Modeling Technique Selection](#4.1) 
   - 4.2 [Test Design](#4.2)
   - 4.3 [Model Building](#4.3)
   - 4.4 [Model Assessment](#4.4)
   
   
5. [Evaluation](#5)
   - 5.1 [Evaluate Results](#5.1)
   - 5.2 [Review Process](#5.2)
   - 5.3 [Next Steps](#5.3)
 
 
6. [Deployment Suggestion](#6)
   - 6.1 [Deployment Plan](#6.1)
   - 6.2 [Plan Monitoring and Maintenance](#6.2)
   - 6.3 [Final Report](#6.3)
   - 6.4 [Project Review](#6.4)

# 1. Business Understanding <a class="anchor" id="1"></a>

## 1.1 Business Objective <a class="anchor" id="1.1"></a>

To better understand customer profiles, especially from the marketing department's perspectives.

## 1.2 Situation Assessment <a class="anchor" id="1.2"></a>

**Computational power**

Computer1: 

Computer2: 

Computer3: 
- CPU: Intel(R) Core i5-8250U CPU
- Memory: 8GB
- GPU: Intel UHD Graphics 620
    
No significant hindrance identified in the initial stage, considering the size of the data and the required task.

**Programming Language:** Python

**Available / utilized libraries**

In [2]:
import pandas as pd
import numpy as np
from math import ceil


# Data Exploration
from pandas_profiling import ProfileReport



#Clustering
from kmodes.kprototypes import KPrototypes
#hierarchical clustering
#from sklearn.cluster import AgglomerativeClustering, KMeans
#from scipy.cluster.hierarchy import dendrogram, linkage, set_link_color_palette

# Missing Values imputation
#from sklearn.impute import KNNImputer
#from sklearn.preprocessing import OrdinalEncoder

#outliers
#from sklearn.ensemble import IsolationForest
#from sklearn.neighbors import LocalOutlierFactor


# Visualization
import matplotlib
matplotlib.use('TkAgg', force=True)
import matplotlib.pyplot as plt
import seaborn as sns
# Import TSNE
#from sklearn.manifold import TSNE



import warnings
warnings.filterwarnings('ignore')

## 1.3 Data Mining Goals  <a class="anchor" id="1.3"></a>

In [3]:
# Conduct customer segmentation on the customer data => data mining problem type is segmentation


## 1.4 Project Plan  <a class="anchor" id="1.4"></a>

In [4]:
# possible clustering techniques: KNN or K-prototype, dependig on the data.

# 2. Data Understanding <a class="anchor" id="2"></a>

## 2.1 Initial Data Collection <a class="anchor" id="2.1"></a>

We have only one source of data hence assessment of potential quality problem from data merge is not applicable

In [5]:
df = pd.read_sas('corrected/a2z_insurance.sas7bdat')
df.tail()

CustID  FirstPolYear  BirthYear         EducDeg  MonthSal  GeoLivArea  \
10291  10292.0        1984.0     1949.0      b'4 - PhD'    3188.0         2.0   
10292  10293.0        1977.0     1952.0    b'1 - Basic'    2431.0         3.0   
10293  10294.0        1994.0     1976.0  b'3 - BSc/MSc'    2918.0         1.0   
10294  10295.0        1981.0     1977.0    b'1 - Basic'    1971.0         2.0   
10295  10296.0        1990.0     1981.0      b'4 - PhD'    2815.0         1.0   

       Children  CustMonVal  ClaimsRate  PremMotor  PremHousehold  PremHealth  \
10291       0.0       -0.11        0.96     393.74          49.45      173.81   
10292       0.0     1405.60        0.00     133.58        1035.75      143.25   
10293       1.0      524.10        0.21     403.63         132.80      142.25   
10294       1.0      250.05        0.65     188.59         211.15      198.37   
10295       1.0      463.75        0.27     414.08          94.45      141.25   

       PremLife  PremWork  
10291      9.78     14.78  
10292     12.89    105.13  
10293     12.67      4.89  
10294     63.90    112.91  
10295      6.89     12.89

## 2.2 Data Description  <a class="anchor" id="2.2"></a>
Based on the metadata provided with some errors in the project description was corrected (eg., all the variable names, description of Birthday was "Customer’s first policy", etc)

Data on 10,290 Customers of the insurance company

| Name | Meaning | Additional Info | 
| :--: | :--------| :--------| 
| CustID | ID | - | 
| FirstPolYear | Year of the customer's first policy | May be considered as the first year as a customer |
| Birthday | Customer's birthday | The current year of the database is 2016|
| EducDeg | Academic Degree | - |
| MonthSal | Gross monthly salary (€) | - |
| GeoLivArea | Living area | No further information provided about the meaning of the area codes |
| Children | Binary variable (Y=1) | - |
| CustMonVal | Customer Monetary Value | Lifetime value = (annual profit from the customer) X (number of years that they are a customer) - (acquisition cost) |
| ClaimsRate | Claims Rate| Amount paid by the insurance company (€)/ Premiums (€) Note: in the last 2 years|
| PremMotor | Premiums (€) in LOB: Motor | Annual Premiums (2016) Negative premiums may manifest reversals occurred in the current year, paid in previous one(s). Applicable to all Prem* variables |
| PremHousehold | Premiums (€) in LOB: Household | * |
| PremHealth | Premiums (€) in LOB: Health | * |
| PremLife | Premiums (€) in LOB: Life | * |
| PremWork | Premiums (€) in LOB: Work| * |

## 2.3 Data Exploration  <a class="anchor" id="2.3"></a>

In [6]:
# Check default import data types of the features
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10296 entries, 0 to 10295
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   CustID         10296 non-null  float64
 1   FirstPolYear   10266 non-null  float64
 2   BirthYear      10279 non-null  float64
 3   EducDeg        10279 non-null  object 
 4   MonthSal       10260 non-null  float64
 5   GeoLivArea     10295 non-null  float64
 6   Children       10275 non-null  float64
 7   CustMonVal     10296 non-null  float64
 8   ClaimsRate     10296 non-null  float64
 9   PremMotor      10262 non-null  float64
 10  PremHousehold  10296 non-null  float64
 11  PremHealth     10253 non-null  float64
 12  PremLife       10192 non-null  float64
 13  PremWork       10210 non-null  float64
dtypes: float64(13), object(1)
memory usage: 1.1+ MB


In [7]:
df.describe()

CustID  FirstPolYear     BirthYear      MonthSal    GeoLivArea  \
count  10296.00000  10266.000000  10279.000000  10260.000000  10295.000000   
mean    5148.50000   1991.062634   1968.007783   2506.667057      2.709859   
std     2972.34352    511.267913     19.709476   1157.449634      1.266291   
min        1.00000   1974.000000   1028.000000    333.000000      1.000000   
25%     2574.75000   1980.000000   1953.000000   1706.000000      1.000000   
50%     5148.50000   1986.000000   1968.000000   2501.500000      3.000000   
75%     7722.25000   1992.000000   1983.000000   3290.250000      4.000000   
max    10296.00000  53784.000000   2001.000000  55215.000000      4.000000   

           Children     CustMonVal    ClaimsRate     PremMotor  PremHousehold  \
count  10275.000000   10296.000000  10296.000000  10262.000000   10296.000000   
mean       0.706764     177.892605      0.742772    300.470252     210.431192   
std        0.455268    1945.811505      2.916964    211.914997     352.595984   
min        0.000000 -165680.420000      0.000000     -4.110000     -75.000000   
25%        0.000000      -9.440000      0.390000    190.590000      49.450000   
50%        1.000000     186.870000      0.720000    298.610000     132.800000   
75%        1.000000     399.777500      0.980000    408.300000     290.050000   
max        1.000000   11875.890000    256.200000  11604.420000   25048.800000   

         PremHealth      PremLife      PremWork  
count  10253.000000  10192.000000  10210.000000  
mean     171.580833     41.855782     41.277514  
std      296.405976     47.480632     51.513572  
min       -2.110000     -7.000000    -12.000000  
25%      111.800000      9.890000     10.670000  
50%      162.810000     25.560000     25.670000  
75%      219.820000     57.790000     56.790000  
max    28272.000000    398.300000   1988.700000

In [8]:
# Check the number of entries.
df.shape

(10296, 14)

In [9]:
# Check for duplicates of the unique customer ID
df.duplicated(subset='CustID').sum()

0

In [10]:
# EducDeg appears to be ordinal. In which case, it would be better to convert it before producing profile report. (Normally would be included in Feature Engineering section)
df.EducDeg.unique()

array([b'2 - High School', b'1 - Basic', b'3 - BSc/MSc', b'4 - PhD', nan],
      dtype=object)

In [11]:
# Decode to handle bytes then extract the numeric characters and replace the string to numeric, ordinal variable.
df['EducDeg'] = df['EducDeg'].str.decode('UTF-8').str.extract('(\d+)', expand=False)
df

CustID  FirstPolYear  BirthYear EducDeg  MonthSal  GeoLivArea  \
0          1.0        1985.0     1982.0       2    2177.0         1.0   
1          2.0        1981.0     1995.0       2     677.0         4.0   
2          3.0        1991.0     1970.0       1    2277.0         3.0   
3          4.0        1990.0     1981.0       3    1099.0         4.0   
4          5.0        1986.0     1973.0       3    1763.0         4.0   
...        ...           ...        ...     ...       ...         ...   
10291  10292.0        1984.0     1949.0       4    3188.0         2.0   
10292  10293.0        1977.0     1952.0       1    2431.0         3.0   
10293  10294.0        1994.0     1976.0       3    2918.0         1.0   
10294  10295.0        1981.0     1977.0       1    1971.0         2.0   
10295  10296.0        1990.0     1981.0       4    2815.0         1.0   

       Children  CustMonVal  ClaimsRate  PremMotor  PremHousehold  PremHealth  \
0           1.0      380.97        0.39     375.85          79.45      146.36   
1           1.0     -131.13        1.12      77.46         416.20      116.69   
2           0.0      504.67        0.28     206.15         224.50      124.58   
3           1.0      -16.99        0.99     182.48          43.35      311.17   
4           1.0       35.23        0.90     338.62          47.80      182.59   
...         ...         ...         ...        ...            ...         ...   
10291       0.0       -0.11        0.96     393.74          49.45      173.81   
10292       0.0     1405.60        0.00     133.58        1035.75      143.25   
10293       1.0      524.10        0.21     403.63         132.80      142.25   
10294       1.0      250.05        0.65     188.59         211.15      198.37   
10295       1.0      463.75        0.27     414.08          94.45      141.25   

       PremLife  PremWork  
0         47.01     16.89  
1        194.48    106.13  
2         86.35     99.02  
3         35.34     28.34  
4         18.78     41.45  
...         ...       ...  
10291      9.78     14.78  
10292     12.89    105.13  
10293     12.67      4.89  
10294     63.90    112.91  
10295      6.89     12.89  

[10296 rows x 14 columns]

In [12]:
# Pandas prifiling - disabled after the initial report 
#profile = ProfileReport(df, title="Pandas Profiling Report")
#profile.to_file("DM_PandasProfile.html")
#profile.to_notebook_iframe()

**Consideration A**

A1. Some decimal values may need to be converted to integers (e.g., year). Need to investigate the nature of those values and required treatment (eg., year 1975.7 can't just be rounded to 1976 without knowing the nature of this entry)<br>
A2. Presence of extreme values that are obviously impossible (fails coherence check) in FirstPolYear and BirthYear.<br>
A3. Outliers that are theoretically possible (or uncertain whether it's impossible), hence requiring further investigation in MonthSal, ClaimsRate, and Prem variables.<br>
A4. EducDeg can be converted to numeric, ordinal variable<br>
A5. GeoLivArea and Children == Categorical (should be converted from float to object type)<br>
A6. Some missing values in most variables<br>
A7. There are 10296 entries instead of 10290 entries as stated in the project description. Quick duplicate check did not reveal any reason behind it. Needs to be confirmed (UPDATE 21 Oct: Consider it a rounding error)<br>


The section below investigates these considerations:

## 2.4 Data Quality Verification  <a class="anchor" id="2.4"></a>

### 2.4.1 Decimal values check  <a class="anchor" id="2.4.1"></a>

In regards to Consideration A1

In [13]:
# Check 
tmp = int(len(df)) # max length

# Instead of counting what is not an integer, we need to count what IS an integer and subtract from the max length due to the presence of NAN values.
for column in df:
    if (df[column].dtype) == "float64":
        only_num = [num for num in df[column] if num % 1 == 0]
        count = tmp-len(only_num)
        print('{}: {}'.format(column, count))  

CustID: 0
FirstPolYear: 30
BirthYear: 17
MonthSal: 36
GeoLivArea: 1
Children: 21
CustMonVal: 9653
ClaimsRate: 9777
PremMotor: 10296
PremHousehold: 9205
PremHealth: 10294
PremLife: 9433
PremWork: 9376


In [14]:
# Compare with null values. The result shows some columns--CustID, FirstPolYear, BirthYear, MonthSal, GeoLivArea, Children-- are only integers mixed with nulls.
df.loc[:, df.columns != 'EducDeg'].isnull().sum()

CustID             0
FirstPolYear      30
BirthYear         17
MonthSal          36
GeoLivArea         1
Children          21
CustMonVal         0
ClaimsRate         0
PremMotor         34
PremHousehold      0
PremHealth        43
PremLife         104
PremWork          86
dtype: int64

### 2.4.2 Initial Coherence Check  <a class="anchor" id="2.4.2"></a>
on FirstPolYear and BirthYear

**FirstPolYear**

In [15]:
# The min value of FirstPolYear (1974) is a possible value. Hence only the largest value needs to be checked.
df.nlargest(10, 'FirstPolYear')

CustID  FirstPolYear  BirthYear EducDeg  MonthSal  GeoLivArea  Children  \
9294  9295.0       53784.0     1948.0       3    3268.0         1.0       0.0   
23      24.0        1998.0     1961.0       2    2870.0         4.0       1.0   
65      66.0        1998.0     1941.0       3    3760.0         1.0       0.0   
279    280.0        1998.0     1938.0       1    4386.0         1.0       0.0   
506    507.0        1998.0     1942.0       1    3743.0         1.0       0.0   
627    628.0        1998.0     1973.0       3    2915.0         1.0       1.0   
734    735.0        1998.0     1945.0       3    3481.0         1.0       0.0   
772    773.0        1998.0     1949.0       2    3601.0         2.0       NaN   
810    811.0        1998.0     1955.0       2    2899.0         4.0       1.0   
816    817.0        1998.0     1990.0       2    1473.0         3.0       1.0   

      CustMonVal  ClaimsRate  PremMotor  PremHousehold  PremHealth  PremLife  \
9294      348.85        0.37     351.40          15.00      187.37     22.45   
23        290.05        0.64     222.04         329.50      248.38     48.01   
65        132.27        0.78     276.83         175.60      261.27      6.89   
279       193.04        0.79      96.02         541.20      189.59    127.69   
506       -46.67        1.02      85.24         554.00      307.39     46.12   
627        42.67        0.90     401.52          89.45      151.14     17.67   
734       101.79        0.81     181.48          89.45      338.73     15.78   
772        27.01        0.94     104.91         308.95      285.61     64.90   
810       682.01        0.41     314.06         737.35       93.35     30.34   
816       478.43        0.45     152.03         402.30      265.05     74.68   

      PremWork  
9294     17.78  
23       33.56  
65       -6.11  
279      90.35  
506      24.45  
627      21.78  
734      53.12  
772      52.90  
810      22.78  
816      23.78

No indication the data integrity is compromised, considering there's only one illogical value (53784). Dropping the row can be considered.

**BirthYear**

In [16]:
# The max value of BirthYear (2001) is possible. Check the mins. 
df.nsmallest(10, 'BirthYear')

CustID  FirstPolYear  BirthYear EducDeg  MonthSal  GeoLivArea  Children  \
7195  7196.0        1993.0     1028.0       3    2830.0         4.0       0.0   
329    330.0        1979.0     1935.0       2    4292.0         1.0       1.0   
958    959.0        1996.0     1935.0       1    4133.0         2.0       1.0   
1013  1014.0        1996.0     1935.0       3    4247.0         2.0       0.0   
3191  3192.0        1990.0     1935.0       3    4475.0         3.0       0.0   
3293  3294.0        1991.0     1935.0       2    3859.0         3.0       0.0   
3454  3455.0        1984.0     1935.0       2    3446.0         3.0       0.0   
7266  7267.0        1975.0     1935.0       4    4597.0         3.0       0.0   
7440  7441.0        1995.0     1935.0       2    3782.0         1.0       0.0   
7956  7957.0        1997.0     1935.0       2    3767.0         2.0       0.0   

      CustMonVal  ClaimsRate  PremMotor  PremHousehold  PremHealth  PremLife  \
7195      146.02        0.77     428.97         192.80      108.91      1.89   
329       -47.77        1.04     457.64         -20.55       94.35     47.23   
958       594.33        0.38     257.05         511.20      129.58     29.56   
1013      -98.79        1.09     142.36         299.50      307.17     24.67   
3191      433.97        0.45     111.80         339.50      268.83    106.91   
3293      644.45        0.25     144.47         358.95      178.59    166.03   
3454      296.41        0.57     201.37         219.50      228.82     51.90   
7266      295.53        0.41     460.53         -65.00      134.58     -5.00   
7440       87.80        0.84     237.60         161.70      262.16     37.34   
7956       68.81        0.85     172.81          87.25      273.83     46.23   

      PremWork  
7195     23.67  
329      -1.11  
958      78.57  
1013     73.79  
3191     12.56  
3293     50.01  
3454     50.01  
7266     14.89  
7440     11.56  
7956     52.79

No indication the data integrity is compromised, considering there's only one illogical value (1028). Dropping the row can be considered or imputation from EducDeg.

In [17]:
# Check for entries where BirthYear < FirstPolYear (Assuming an unborn person cannot be insured)
df.query('FirstPolYear < BirthYear')

CustID  FirstPolYear  BirthYear EducDeg  MonthSal  GeoLivArea  \
1          2.0        1981.0     1995.0       2     677.0         4.0   
13        14.0        1983.0     2000.0       1    1043.0         3.0   
18        19.0        1981.0     1982.0       1    1414.0         3.0   
22        23.0        1976.0     1986.0       2    1495.0         1.0   
40        41.0        1994.0     1995.0       2    1177.0         4.0   
...        ...           ...        ...     ...       ...         ...   
10268  10269.0        1987.0     1988.0       3    1414.0         3.0   
10273  10274.0        1989.0     1992.0       2    1035.0         1.0   
10284  10285.0        1980.0     1987.0       3    1504.0         4.0   
10288  10289.0        1982.0     1993.0       2    1465.0         1.0   
10290  10291.0        1994.0     1999.0       1     626.0         3.0   

       Children  CustMonVal  ClaimsRate  PremMotor  PremHousehold  PremHealth  \
1           1.0     -131.13        1.12      77.46         416.20      116.69   
13          1.0      -75.12        1.06      44.34         342.85      127.69   
18          1.0      230.38        0.71     313.17         373.40      121.80   
22          1.0      -89.79        1.08     209.04         308.95      192.48   
40          1.0      121.36        0.84      52.01         455.65      135.47   
...         ...         ...         ...        ...            ...         ...   
10268       1.0       50.46        0.90     129.58         167.80      184.48   
10273       1.0      293.51        0.58     235.71         195.60      103.13   
10284       1.0       -1.55        0.96     390.63          29.45      179.70   
10288       1.0      795.15        0.35      67.79         820.15      102.13   
10290       1.0      176.26        0.85       6.89         878.50      103.13   

       PremLife  PremWork  
1        194.48    106.13  
13       267.94     94.46  
18        14.78     72.57  
22        64.79     54.90  
40       146.36    148.03  
...         ...       ...  
10268    165.92     75.46  
10273    170.92     58.12  
10284     -6.00       NaN  
10288    182.48     86.46  
10290    113.02    201.26  

[1997 rows x 14 columns]

1997 rows = 19% of the total data. Data integrity may have been compromised, requires a clarification. Considering the originally incorrect description of BirthYear, this feature may be the problem rather than FirstPolYear.

In [18]:
# Deviation check

In [19]:
df.nlargest(10, 'MonthSal')

CustID  FirstPolYear  BirthYear EducDeg  MonthSal  GeoLivArea  Children  \
5882  5883.0        1995.0     1956.0       2   55215.0         4.0       1.0   
8261  8262.0        1981.0     1991.0       2   34490.0         4.0       0.0   
7510  7511.0        1979.0     1937.0       2    5021.0         1.0       0.0   
4244  4245.0        1996.0     1936.0       3    4995.0         2.0       0.0   
4440  4441.0        1986.0     1940.0       3    4904.0         2.0       0.0   
6902  6903.0        1989.0     1941.0       3    4897.0         4.0       0.0   
4772  4773.0        1984.0     1940.0       2    4883.0         4.0       0.0   
8284  8285.0        1995.0     1941.0       3    4883.0         3.0       1.0   
9433  9434.0        1984.0     1940.0       2    4883.0         4.0       0.0   
1406  1407.0        1993.0     1941.0       1    4872.0         4.0       0.0   

      CustMonVal  ClaimsRate  PremMotor  PremHousehold  PremHealth  PremLife  \
5882      122.25        0.79     281.83         147.25      130.58     84.46   
8261      608.89        0.28      57.01         358.95      195.26    113.80   
7510      -25.00        1.00     237.71         241.15      237.60     40.34   
4244      338.40        0.60     308.39         368.40      175.70     30.56   
4440      363.85        0.50     205.26         216.15      267.94     60.01   
6902       19.34        0.93     328.73          64.45      196.26     18.78   
4772      -15.11        0.99     399.52         221.70      101.02      9.89   
8284      444.54        0.30     337.62         117.25      164.92     18.67   
9433       20.12        0.94     399.52         221.70      101.02      9.89   
1406      246.61        0.69     212.93         391.75      220.04     19.56   

      PremWork  
5882     66.68  
8261    161.14  
7510     49.12  
4244     25.56  
4440     32.56  
6902     44.34  
4772       NaN  
8284     33.34  
9433     35.23  
1406     39.12

index 5882 and 8261 => outliers with values 55215 and 34490

In [20]:
df.nlargest(10, 'CustMonVal')

CustID  FirstPolYear  BirthYear EducDeg  MonthSal  GeoLivArea  Children  \
5293  5294.0        1989.0     1963.0       3    3564.0         3.0       1.0   
7085  7086.0        1992.0     1963.0       3    2460.0         3.0       1.0   
6701  6702.0        1995.0     1951.0       3    2947.0         4.0       0.0   
5084  5085.0        1994.0     1959.0       2    3330.0         3.0       0.0   
2323  2324.0        1984.0     1997.0       1     782.0         3.0       1.0   
8667  8668.0        1981.0     1997.0       1    1421.0         1.0       0.0   
8366  8367.0        1981.0     1995.0       1    1094.0         1.0       1.0   
4373  4374.0        1976.0     1993.0       1    1133.0         4.0       1.0   
2043  2044.0        1995.0     1996.0       1    1089.0         3.0       0.0   
9235  9236.0        1997.0     1998.0       1     566.0         4.0       1.0   

      CustMonVal  ClaimsRate  PremMotor  PremHousehold  PremHealth  PremLife  \
5293    11875.89        0.09   11604.42          48.90     1045.52    103.13   
7085     5596.84        0.01    5645.50         -25.55       49.01      1.89   
6701     4328.50        0.03    4273.49          83.90      105.13      3.00   
5084     2314.21        0.14     350.51         201.70      125.80     39.23   
2323     2054.07        0.03      39.12        1957.60       47.23     15.78   
8667     1997.60        0.05      30.34        1924.25       37.23     33.34   
8366     1891.04        0.09      13.67        1918.15       51.90     15.78   
4373     1801.45        0.10      53.01        1826.45       68.68      3.89   
2043     1716.00        0.11      67.90        1673.10       65.90    112.02   
9235     1691.43        0.15      14.56        1777.55       49.23     46.01   

      PremWork  
5293    296.47  
7085      1.89  
6701      3.89  
5084   1988.70  
2323     82.35  
8667     98.35  
8366     99.02  
4373     88.13  
2043     27.56  
9235    121.69

In [21]:
df.nsmallest(20, 'CustMonVal')

CustID  FirstPolYear  BirthYear EducDeg  MonthSal  GeoLivArea  Children  \
171    172.0        1990.0     1989.0       3    1086.0         4.0       1.0   
655    656.0        1974.0     1963.0       3    3279.0         3.0       1.0   
675    676.0        1994.0     1937.0       2    4435.0         1.0       0.0   
9648  9649.0        1990.0     1950.0       3    3574.0         3.0       0.0   
707    708.0        1991.0     1991.0       3    1634.0         1.0       1.0   
917    918.0        1986.0     1952.0       2    3355.0         4.0       0.0   
51      52.0        1982.0     1966.0       3    3234.0         4.0       1.0   
135    136.0        1997.0     1988.0       3    2176.0         4.0       1.0   
758    759.0        1988.0     1992.0       3    1117.0         1.0       1.0   
112    113.0        1997.0     1979.0       4    2354.0         1.0       1.0   
793    794.0        1980.0     1983.0       3    1370.0         3.0       1.0   
982    983.0        1989.0     1981.0       2    1771.0         3.0       1.0   
951    952.0        1998.0     1976.0       3    1586.0         3.0       1.0   
1025  1026.0        1988.0     1939.0       3    4566.0         1.0       1.0   
9390  9391.0        1992.0     1999.0       1     540.0         3.0       0.0   
9954  9955.0        1974.0     1998.0       1     768.0         4.0       1.0   
895    896.0        1988.0     1994.0       1    1166.0         1.0       0.0   
695    696.0        1987.0     1998.0       2    1219.0         1.0       1.0   
5815  5816.0        1990.0     1996.0       1    1120.0         4.0       1.0   
8659  8660.0        1977.0     1996.0       1     693.0         3.0       1.0   

      CustMonVal  ClaimsRate  PremMotor  PremHousehold  PremHealth  PremLife  \
171   -165680.42      256.20     378.07          78.90      166.81      6.89   
655    -64891.00       96.00     410.30         117.25       95.35     37.34   
675    -52382.76       63.00     197.48         280.60      276.94     51.12   
9648   -37327.08       69.00     319.06         -25.55      245.38      4.89   
707    -28945.40       35.00     175.70         319.50      294.39     44.12   
917    -26130.45       32.30     135.58         290.05      221.82     86.46   
51     -14714.08       25.36     557.44          20.00       29.56      5.00   
135    -10198.91       14.80     297.61         162.80      143.36    136.47   
758    -10107.37       12.60     193.37         342.85      276.94      8.78   
112     -8719.04       15.65     518.32           4.45       55.90      3.89   
793     -7851.17       13.90     127.58          48.35      398.41     19.56   
982     -6115.85        9.87      83.35          98.35      336.84     74.68   
951     -2642.91        5.60     501.65         -20.00       90.46     -1.00   
1025    -2082.83        4.36     370.07          14.45      189.59     14.78   
9390     -490.20        1.55      20.78         290.60       47.12    299.50   
9954     -416.73        1.55      21.67         136.70       27.45    177.81   
895      -406.07        1.50      11.67         225.60      125.58    143.25   
695      -402.96        1.47      57.12         236.15       79.57    193.48   
5815     -386.18        1.42      39.23         348.95       79.57    191.48   
8659     -382.40        1.47     149.25         202.25       57.90    255.16   

      PremWork  
171      18.45  
655      22.56  
675      38.34  
9648      4.78  
707      16.89  
917     100.13  
51       -9.00  
135      -3.00  
758      47.23  
112      10.89  
793      12.78  
982      93.46  
951      -2.00  
1025     23.56  
9390    186.48  
9954    350.62  
895     249.49  
695     241.60  
5815    208.93  
8659    102.24

In [22]:
# Lucas: does claim rate > 100 make sense?

### 2.4.3 Outliers
Concept of "outlier" doesn't really apply to categorical variables.

In [23]:
plt.get_backend()

'TkAgg'

In [24]:
metric_features = [v for v in df.describe().columns if v not in ['CustID','GeoLivArea']]

In [25]:
# All Numeric Variables' Box Plots in one figure
sns.set()

# Prepare figure. Create individual axes where each box plot will be placed
fig, axes = plt.subplots(2, ceil(len(metric_features) / 2), figsize=(20, 11))

# Plot data
# Iterate across axes objects and associate each box plot (hint: use the ax argument):
for ax, feat in zip(axes.flatten(), metric_features): # Notice the zip() function and flatten() method
    sns.boxplot(x=df[feat], ax=ax)
    
# Layout
# Add a centered title to the figure:
title = "Numeric Variables' Box Plots"

plt.suptitle(title)

#plt.savefig(os.path.join('..', 'figures', 'numeric_variables_boxplots.png'), dpi=200)
plt.show()

In [26]:
# Lucas: These are arbitrary thresholds and we are not sure yet whether we just wanna drop these rows or do something else with it
# so I disabled it

#df.loc[df['FirstPolYear']==53784]
#df.loc[df['BirthYear']<1900]
#df.loc[df['MonthSal']>20000]
#df.loc[df['CustMonVal']<-50000]
#df.loc[df['ClaimsRate']>50]
#df.loc[df['PremMotor']>2500]
#df.loc[df['PremHousehold']>3000]
#df.loc[df['PremHealth']>3000]
#df.loc[df['PremWork']>500]
#new_df = df.drop([9294,7195,5882,8261,171,655,675,9648,5128,5211,5293,6701,7085,7904,6343,6835,8866,7961,9149,5084,7961])
#

In [ ]:
# Pairwise Relationship of Numerical Variables
sns.set()

# Setting pairplot
sns.pairplot(df[metric_features], diag_kind="hist",hue='Children') # new_df variable removed for now

# Layout
plt.subplots_adjust(top=0.95)
plt.suptitle("Pairwise Relationship of Numerical Variables", fontsize=20)

plt.show()

### 2.4.x Inconsistency in the number of customers
Related to Consideration A7

In [ ]:
# The number of rows exceeds the number of customers by 6. Check for duplicates excluding CustID (unique)
tmp = df.loc[:, df.columns != 'CustID']
tmp.drop_duplicates()
tmp

It appears to be a rounding error

# 3. Data Preparation  <a class="anchor" id="3"></a>

## 3.1 Data Selection  <a class="anchor" id="3.1"></a>

We are tasked with a segmentation problem, without specific feature of interest communicated from the insurance company. Therefore feature selection can be done qualitatively, with some consideration of results found from the previous sections (e.g., multicollinearity)

we can mention Lucas's job as a industry insider's opinion?

## 3.2 Data Cleaning  <a class="anchor" id="3.2"></a>

**missing values**

**outliers**

## 3.3 Data Construction  <a class="anchor" id="3.3"></a>

In [ ]:
#age from birthyear, 

## 3.4 Data Integration  <a class="anchor" id="3.4"></a>

## 3.5 Data Formatting  <a class="anchor" id="3.5"></a>

In [ ]:
# GeoLiveArea,  Children = Object

## 3.6 Data Description  <a class="anchor" id="3.6"></a>